NitroxClient/Debuggers/SceneDebugger.cs

```diff
@@ -0,0 +1,234 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using NitroxClient.Unity.Helper;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+namespace NitroxClient.Debuggers
+{
+    public class SceneDebugger : BaseDebugger
+    {
+        private Vector2 gameObjectScrollPos;
+        private Vector2 hierarchyScrollPos;
+        private GameObject selectedObject;
+        private Scene selectedScene;
+
+        public SceneDebugger() : base(400, null, KeyCode.S, true, false, false, GUISkinCreationOptions.DERIVEDCOPY)
+        {
+            Tabs.AddRange(new[]
+            {
+                "Scenes",
+                "Hierarchy",
+                "GameObject"
+            });
+        }
+
+        private enum TabTypes
+        {
+            SCENES,
+            HIERARCHY,
+            GAMEOBJECT
+        }
+
+        protected override void OnSetSkin(GUISkin skin)
+        {
+            base.OnSetSkin(skin);
+
+            skin.SetCustomStyle("sceneLoaded", skin.label, s =>
+            {
+                s.normal = new GUIStyleState()
+                {
+                    textColor = Color.green
+                };
+                s.fontStyle = FontStyle.Bold;
+            });
+
+            skin.SetCustomStyle("loadScene", skin.button, s =>
+            {
+                s.fixedWidth = 60;
+            });
+
+            skin.SetCustomStyle("mainScene", skin.label, s =>
+            {
+                s.margin = new RectOffset(2, 10, 10, 10);
+                s.alignment = TextAnchor.MiddleLeft;
+                s.fontSize = 20;
+            });
+
+            skin.SetCustomStyle("fillMessage", skin.label, s =>
+            {
+                s.stretchWidth = true;
+                s.stretchHeight = true;
+                s.fontSize = 24;
+                s.alignment = TextAnchor.MiddleCenter;
+                s.fontStyle = FontStyle.Italic;
+            });
+
+            skin.SetCustomStyle("breadcrumb", skin.label, s =>
+            {
+                s.fontSize = 20;
+                s.fontStyle = FontStyle.Bold;
+            });
+
+            skin.SetCustomStyle("breadcrumbNav", skin.box, s =>
+            {
+                s.stretchWidth = false;
+                s.fixedWidth = 100;
+            });
+        }
+
+        protected override void Render(int windowId)
+        {
+            switch ((TabTypes)ActiveTab)
+            {
+                default:
+                case TabTypes.SCENES:
```

Comment:
Do you think this would be more readable if the tab render resolution was included in the base class?  I was thinking that you could have a base class method called AddTab that takes a name and a function pointer to the render function.   This could simply add it to a map in the base class that can be looked up in the render function.  Some benefits of doing this:

1) No need to override render in all of the children
2) No need to declare a new enum in all of the children
3) Allows us to break up some of these really huge switch statements
4) Allows better information hiding by not bubbling up more fields to children classes 

Feel free to take or leave this suggestion.  Either way, I would recommend breaking the cases into methods for readability.

Reply:
This is a very good suggestion. I've changed it and it's even more clear than before. Thanks.